lib/zendesk_apps_support/validations/secrets.rb

```diff
@@ -0,0 +1,76 @@
+        'Scarlett AWS Secret Key' => /([sS][cC][aA][rR][lL][eE][tT][tT][_-][aA][wW][sS][_-][sS][eE][cC][rR][eE][tT][_-][kK][eE][yY].*.[0-9a-zA-Z+.]{35,45})/,
+        'Braintree Key' => /(braintree_key.*.[0-9a-zA-Z]{16,36})/,
+        'Ticket Validation Key' => /(ticket_validation_key.*.[0-9a-zA-Z]{15,25})/,
+        'App Key' => /([aA][pP][pP][-_][kK][eE][yY]( [:\"=-]|[:\"=-]).*.[0-9a-zA-Z+_.-]{4,156})/,
+        'App Secret' => /([aA][pP][pP][-_][sS][eE][cC][rR][eE][tT]( [:\"=-]|[:\"=-]).*.[0-9a-zA-Z+_.-]{4,156})/,
+        'Consumer Key' => /([cC][oO][nN][sS][uU][mM][eE][rR][-_][kK][eE][yY]( [:\"=-]|[:\"=-]).*.[0-9a-zA-Z+_.-]{4,156})/,
+        'Consumer Secret' => /([cC][oO][nN][sS][uU][mM][eE][rR][-_][sS][eE][cC][rR][eE][tT]( [:\"=-]|[:\"=-]).*.[0-9a-zA-Z+_.-]{4,156})/,
+        'Generic Secret' => /(?m)^([sS][eE][cC][rR][eE][tT]( [:\"=-]|[:\"=-]).*.[0-9a-zA-Z+_.-]{4,156})/,
+        'Master Key' => /([mM][aA][sS][tT][eE][rR][-_][kK][eE][yY]( [:\"=-]|[:\"=-]).*.[0-9a-zA-Z+_.-]{4,156})/,
+        'Master Secret' => /([mM][aA][sS][tT][eE][rR][-_][sS][eE][cC][rR][eE][tT]( [:\"=-]|[:\"=-]).*.[0-9a-zA-Z+_.-]{4,156})/,
+        'Token Key' => /([tT][oO][kK][eE][nN][-_][kK][eE][yY]( [:\"=-]|[:\"=-]).*.[0-9a-zA-Z+_.-]{4,156})/,
+        'Token Secret' => /([tT][oO][kK][eE][nN][-_][sS][eE][cC][rR][eE][tT]( [:\"=-]|[:\"=-]).*.[0-9a-zA-Z+_.-]{4,156})/,
+        'Zendesk Zopim Mobile SSO Key' => /(zendesk_zopim_mobile_sso_key.*.[0-9a-f]{58,68})/,
+        'Help Center Private Key' => /([pP][rR][iI][vV][aA][tT][eE][-_][kK][eE][yY]( [:\"=-]|[:\"=-]).*.[0-9a-zA-Z+_.-]{4,156})/,
+        'X-Outbound-Key' => /([xX][-][oO][uU][tT][bB][oO][uU][nN][dD][-][kK][eE][yY][:\" \t=-].*.[0-9a-z-]{32,36})/,
+        'Attachment Token Key' => /(attachment_token_key.*.[0-9a-f]{24,72})/,
+        'Password' => /([pP][aA][sS][sS][wW][oO][rR][dD].*.[0-9a-zA-Z+_.-]{4,156})/,
+        'Token' => /([tT][oO][kK][eE][nN]( [:\"=-]|[:\"=-]).*.[0-9a-zA-Z+_.-]{4,156})/
+        # rubocop:enable Metrics/LineLength
+      }.freeze
+
+      class << self
+        def call(package)
+          compromised_files = package.text_files.map do |file|
+            contents = file.read
+
+            APPLICATION_SECRETS.each do |secret_type, regex_str|
+              next unless contents =~ Regexp.new(regex_str)
+              package.warnings << I18n.t('txt.apps.admin.warning.app_build.application_secret',
+                                         file: file.relative_path,
+                                         secret_type: secret_type)
+            end
+
+            file.relative_path if contents =~ Regexp.union(SECRET_KEYWORDS)
+          end.compact
+
+          if compromised_files.any?
+            package.warnings << I18n.t('txt.apps.admin.warning.app_build.generic_secrets',
+                                       files: compromised_files.join(', '),
+                                       count: compromised_files.count)
+          end
+          []
```

Comment:
Is there a reason why we return an empty array here? Should we be returning package or package.warning?

Reply:
In the test, we assign

```
errors = subject.call(package)
```

Since the method returns an empty array, does it mean that expect(errors).to be_empty will always be the case?

Reply:
Ah yeah so admittedly this isn't super clear - in https://github.com/zendesk/zendesk_apps_support/blob/master/lib/zendesk_apps_support/package.rb#L31, we collate errors from each of these "validator" classes. Each validator returns an array of errors, but since this one only displays warnings (no 'blocking' errors) I just set the return value to an empty array here.

I could instead assign `errors = []` at the start of this file (e.g. https://github.com/zendesk/zendesk_apps_support/blob/master/lib/zendesk_apps_support/validations/svg.rb#L56) and return `errors` at the end if you think that's better?

Reply:
If we're never appending to errors I think it could be misleading to have that variable. What about adding a comment `[] // no errors` or something?

Reply:
We actually do something similar in the `banner` validator (https://github.com/zendesk/zendesk_apps_support/blob/master/lib/zendesk_apps_support/validations/banner.rb#L30), but it is kind of confusing.

Reply:
As discussed in person, we decided to perhaps call the method on its own. eg

`errors << validate_no_secrets()` vs `validate_no_secrets()`

as `package.warnings` needs not append errors collection